### PR TITLE
Guitar/app acc

### PR DESCRIPTION
**Pull request by guitarsheng**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [x] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [x] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [x] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
The precondition is to put your code inside this repository and upgrade the version in the pull request but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
spring

File: src/spring/azext_spring/_tanzu_component.py [hunk 1]
```diff
@@ -10,6 +10,7 @@
 from msrestazure.tools import resource_id
 
 from .vendored_sdks.appplatform.v2022_01_01_preview import models
+from .vendored_sdks.appplatform.v2022_11_01_preview import models
```
Comment (yuwzho) [inline src/spring/azext_spring/_tanzu_component.py:13]:
```suggestion
from .vendored_sdks.appplatform.v2022_11_01_preview import models
```

File: src/spring/azext_spring/_tanzu_component.py [hunk 2]
```diff
@@ -61,3 +62,14 @@ def create_api_portal(cmd, client, resource_group, service, enable_api_portal, a
             api_portal_resource.sku = models.Sku(name=sku.name, tier=sku.tier,
                                                  capacity=api_portal_instance_count)
         return client.api_portals.begin_create_or_update(resource_group, service, DEFAULT_NAME, api_portal_resource)
+
+def create_application_accelerator(cmd, client, resource_group, service, enable_application_accelerator, sku=None, **_):
```
Comment (yuwzho) [inline src/spring/azext_spring/_tanzu_component.py:66]:
```suggestion
def create_application_accelerator(cmd, client, resource_group, service, enable_application_accelerator, **_):
```

File: src/spring/azext_spring/_transformers.py [hunk 3]
```diff
@@ -205,3 +205,60 @@ def transform_api_portal_output(result):
...
+            instance_count = component['resourceRequests']['instanceCount']
+            instances = component['instances'] or []
+            running_number = len(
+                [x for x in instances if x['status'].upper() == "RUNNING"])
+            new_entry['Running Instance'] = "{}/{}".format(running_number, instance_count)
+            new_entry['CPU'] = component['resourceRequests']['cpu']
+            new_entry['Memory'] = component['resourceRequests']['memory']
+            new_result.append(new_entry)
+
+    return result if is_list else new_result
```
Comment (yuwzho) [inline src/spring/azext_spring/_transformers.py:233]: This method is used to show the table output for your command, everyline represents one item. You can obtain the output in `az spring xxxxxxx -o table`

File: src/spring/azext_spring/commands.py [hunk 4]
```diff
@@ -92,6 +95,21 @@ def load_command_table(self, _):
...
+
+    predefined_accelerator_cmd_group = CliCommandType(
+        operations_tmpl='azext_spring.predefined_accelerator#{}',
+        client_factory=cf_spring_20221101preview
+    )
+
+    customized_accelerator_cmd_group = CliCommandType(
+        operations_tmpl='azext_spring.customized_accelerator#{}',
+        client_factory=cf_spring_20221101preview
+    )
```
Comment (yuwzho) [inline src/spring/azext_spring/commands.py:111]: This command group is used to specify API version. I encourgage to use the same API version for acc feature. Here you can only keep L98-L101

File: src/spring/azext_spring/commands.py [hunk 5]
```diff
@@ -303,6 +321,30 @@ def load_command_table(self, _):
         g.custom_command('unbind', 'api_portal_custom_domain_unbind')
         g.custom_command('update', 'api_portal_custom_domain_update')
 
+    with self.command_group('spring application-accelerator',
+                            custom_command_type=application_accelerator_cmd_group,
+                            exception_handler=handle_asc_exception) as g:
+        g.custom_show_command('show', 'application_accelerator_show', table_transformer=transform_application_accelerator_output)
+        g.custom_command('create', 'application_accelerator_create', table_transformer=transform_application_accelerator_output)
```
Comment (yuwzho) [inline src/spring/azext_spring/commands.py:328]:
```suggestion
        g.custom_command('create', 'application_accelerator_create', support_no_wait=True, table_transformer=transform_application_accelerator_output)
```

File: src/spring/azext_spring/commands.py [hunk 6]
```diff
@@ -303,6 +321,30 @@ def load_command_table(self, _):
...
+        g.custom_command('delete', 'application_accelerator_delete', supports_no_wait=True)
+
+    with self.command_group('spring application-accelerator predefined-accelerator',
+                            custom_command_type=predefined_accelerator_cmd_group,
+                            exception_handler=handle_asc_exception) as g:
+        g.custom_command('list', 'predefined_accelerator_list', table_transformer=transform_predefined_accelerator_output)                        
+        g.custom_show_command('show', 'predefined_accelerator_show', table_transformer=transform_predefined_accelerator_output)
+        g.custom_command('disable', 'predefined_accelerator_disable', supports_no_wait=True)
+        g.custom_command('enable', 'predefined_accelerator_enable', supports_no_wait=True)
```
Comment (yuwzho) [inline src/spring/azext_spring/commands.py:337]: Can we use create/delete here instead of enbale/disable?

Comment (yuwzho) [inline src/spring/azext_spring/commands.py:337]:
```suggestion
        g.custom_command('disable', 'predefined_accelerator_disable', supports_no_wait=True)
        g.custom_command('enable', 'predefined_accelerator_enable', supports_no_wait=True, table_transformer=transform_predefined_accelerator_output)
```

File: src/spring/azext_spring/commands.py [hunk 7]
```diff
@@ -303,6 +321,30 @@ def load_command_table(self, _):
...
+
+    with self.command_group('spring application-accelerator customized-accelerator',
+                            custom_command_type=customized_accelerator_cmd_group,
+                            exception_handler=handle_asc_exception) as g:
+        g.custom_command('list', 'customized_accelerator_list', table_transformer=transform_customized_accelerator_output)                        
+        g.custom_show_command('show', 'customized_accelerator_show', table_transformer=transform_customized_accelerator_output)
+        g.custom_command('add', 'customized_accelerator_upsert', supports_no_wait=True)
```
Comment (yuwzho) [inline src/spring/azext_spring/commands.py:344]: add or create?

Comment (yuwzho) [inline src/spring/azext_spring/commands.py:344]:
```suggestion
        g.custom_command('add', 'customized_accelerator_upsert', supports_no_wait=True, table_transformer=transform_customized_accelerator_output)
```

File: src/spring/azext_spring/_params.py [hunk 8]
```diff
@@ -761,3 +765,38 @@ def prepare_logs_argument(c):
...
+        with self.argument_context(scope) as c:
+            c.argument('name', name_type, help='Name for customized accelerator.')
+
+    for scope in ['spring application-accelerator customized-accelerator add',
+                  'spring application-accelerator customized-accelerator update']:
+        with self.argument_context(scope) as c:
+            c.argument('display_name', type=str, help='Display name for customized accelerator.')
+            c.argument('description', type=str, help='Description for customized accelerator.')
+            c.argument('icon_url', type=str, help='Icon url for customized accelerator.')
+            c.argument('accelerator_tags', type=str, help='Tags for customized accelerator.')
```
Comment (yuwzho) [inline src/spring/azext_spring/_params.py:790]: comma seprate list?

Comment (yuwzho) [reply src/spring/azext_spring/_params.py:790]: see gateway sso scopes for example

File: src/spring/azext_spring/commands.py (hunk 7 above)
Comment (guitarsheng) [reply src/spring/azext_spring/commands.py:344]: Updated

**Comment by yuwzho** [comment]
#5532

File: src/spring/azext_spring/commands.py (hunk 6 above)
Comment (guitarsheng) [reply src/spring/azext_spring/commands.py:366]: No

File: src/spring/azext_spring/commands.py [hunk 9]
```diff
@@ -332,6 +340,30 @@ def load_command_table(self, _):
         g.custom_command('unbind', 'api_portal_custom_domain_unbind')
         g.custom_command('update', 'api_portal_custom_domain_update')
 
+    with self.command_group('spring application-accelerator',
+                            custom_command_type=application_accelerator_cmd_group,
+                            exception_handler=handle_asc_exception) as g:
```
Comment (yuwzho) [inline src/spring/azext_spring/commands.py:345]:
```suggestion
    with self.command_group('spring application-accelerator',
                            custom_command_type=application_accelerator_cmd_group,
                            is_preview=True,
                            exception_handler=handle_asc_exception) as g:
```

Comment (guitarsheng) [reply src/spring/azext_spring/commands.py]: Updated

**Comment by guitarsheng** [comment]
hi @zhoxing-ms 

All checks have passed. It's part of dev tools feature which is relevant to PR as shown in link below 
https://github.com/Azure/azure-cli-extensions/pull/5522

Can you please help to review?

**Review by yuwzho — APPROVED** [review]

File: src/spring/azext_spring/commands.py [hunk 10]
```diff
@@ -332,6 +340,31 @@ def load_command_table(self, _):
...
+        g.custom_command('enable', 'predefined_accelerator_enable', supports_no_wait=True)
+
+    with self.command_group('spring application-accelerator customized-accelerator',
+                            custom_command_type=application_accelerator_cmd_group,
+                            exception_handler=handle_asc_exception) as g:
+        g.custom_command('list', 'customized_accelerator_list', table_transformer=transform_customized_accelerator_output)
+        g.custom_show_command('show', 'customized_accelerator_show', table_transformer=transform_customized_accelerator_output)
+        g.custom_command('create', 'customized_accelerator_upsert', supports_no_wait=True, validator=validate_customized_accelerator)
+        g.custom_command('update', 'customized_accelerator_upsert', supports_no_wait=True, validator=validate_customized_accelerator)
+        g.custom_command('delete', 'customized_accelerator_delete', supports_no_wait=True)
```
Comment (zhoxing-ms) [inline src/spring/azext_spring/commands.py:366]: Do we need to consider adding `confirmation=true` for those `delete` command?

File: src/spring/azext_spring/tests/latest/test_asa_customized_accelerator.py [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+import os
+
+from azure.cli.testsdk import (ScenarioTest, record_only)
+
+# pylint: disable=line-too-long
+# pylint: disable=too-many-lines
+
+
+@record_only()
```
Comment (zhoxing-ms) [inline src/spring/azext_spring/tests/latest/test_asa_customized_accelerator.py:14]: May I ask why those new tests are marked as `record_only()`? Could you add the reason to the comment to explain it?

File: src/spring/azext_spring/tests/latest/test_asa_predefined_accelerator.py [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+
+import os
+
+from azure.cli.testsdk import (ScenarioTest, record_only)
+
+# pylint: disable=line-too-long
+# pylint: disable=too-many-lines
+
+
+@record_only()
```
Comment (zhoxing-ms) [inline src/spring/azext_spring/tests/latest/test_asa_predefined_accelerator.py:14]: Same as above

File: src/spring/azext_spring/commands.py (hunk 10 above)
Comment (guitarsheng) [reply src/spring/azext_spring/commands.py:366]: Updated. Add it to application accelerator delete command, as it will delete all of its sub resources.

**Review by zhoxing-ms — APPROVED** [review]

**Pull request merged**